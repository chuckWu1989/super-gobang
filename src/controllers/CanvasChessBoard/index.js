const drawLine = (ctx, s, e) => {
  ctx.beginPath();
  ctx.moveTo(s[0], s[1]);
  ctx.lineTo(e[0], e[1]);
  ctx.closePath();
  ctx.stroke();
};
export const drawBoard = (canvas, props) => {
  const {
    grid,
    coord,
    width,
    ratio,
  } = props;
  const ctx = canvas.getContext('2d');
  const height = width * ratio;
  ctx.clearRect(0, 0, width, height);
  for (let i = 0; i <= (grid + 1); i += 1) {
    drawLine(ctx, coord[0][i], coord[grid + 1][i]);
    drawLine(ctx, coord[i][0], coord[i][grid + 1]);
  }
};

export default drawBoard;
