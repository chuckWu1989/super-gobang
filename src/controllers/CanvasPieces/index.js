import { radiusRatio, s, e } from './config';

const clearCanvas = (ctx, width, height) => {
  ctx.clearRect(0, 0, width, height);
};
const clearPiece = (ctx, coord, unitWidth) => {
  const cx = coord[0];
  const cy = coord[1];
  ctx.clearRect(cx - (unitWidth / 2), cy - (unitWidth / 2), unitWidth, unitWidth);
};
const drawPiece = (ctx, coord, unitWidth, color) => {
  const r = radiusRatio * (unitWidth / 2);
  const cx = coord[0];
  const cy = coord[1];
  ctx.beginPath();
  ctx.fillStyle = color;
  ctx.arc(cx, cy, r, s, e);
  ctx.fill();
  ctx.closePath();
};
const addOrRemovePiece = (ctx, nextProps, props) => {
  const { chessMoves: nextChessMoves, unitWidth, coord } = nextProps;
  const { chessMoves } = props;
  if (chessMoves.length < nextChessMoves.length) {
    const { x, y, color } = nextChessMoves[nextChessMoves.length - 1];
    const point = coord[x][y];
    drawPiece(ctx, point, unitWidth, color);
  } else if (chessMoves.length > nextChessMoves.length) {
    const { x, y } = chessMoves[chessMoves.length - 1];
    const point = coord[x][y];
    clearPiece(ctx, point, unitWidth);
  }
};
const redrawPieces = (ctx, props) => {
  const {
    width,
    height,
    unitWidth,
    coord,
  } = props;
  const { chessMoves } = props;
  clearCanvas(ctx, width, height);
  chessMoves.forEach((element) => {
    const { x, y, color } = element;
    const point = coord[x][y];
    drawPiece(ctx, point, unitWidth, color);
  });
};
export const detectChessMovesChange = (container, nextProps, props) => {
  const ctx = container.getContext('2d');
  const { chessMoves } = props;
  const { chessMoves: nextChessMoves } = nextProps;
  if (chessMoves.length !== nextChessMoves.length) {
    addOrRemovePiece(ctx, nextProps, props);
  }
};
export const detectWidthChange = (container, prevProps, props) => {
  const ctx = container.getContext('2d');
  const { width } = props;
  const { width: nextWidth } = prevProps;
  if (nextWidth !== width) {
    redrawPieces(ctx, prevProps);
  }
};

export default detectChessMovesChange;
