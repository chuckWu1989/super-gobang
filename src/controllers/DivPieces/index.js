import React from 'react';
import DivPiece from '../../views/DivPiece';
import { radiusRatio } from './config';

/**
 * Clear all content on chessboard
 * @param {array} children The children of component
 * @param {number} s Start point to be sliced
 * @param {number} e End point to be sliced
 * @returns {array} Updated children
 */
export function clearPiece(children, s, e) {
  const nextChildren = children.slice(s, e);
  return nextChildren;
}
/**
 * Draw piece on chessboard
 * @param {array} children The children of component
 * @param {array} coord The coordinate of piece
 * @param {number} unitWidth Unit width per grid
 * @param {string} color The color of piece
 * @returns {array} Updated children
 */
export const drawPiece = (children, coord, unitWidth, color) => {
  const size = radiusRatio * unitWidth;
  const x = (coord[0] - (size / 2)) + 1;
  const y = (coord[1] - (size / 2)) + 1;
  const key = `${x}-${y}-piece`;
  const piece = <DivPiece key={key} x={x} y={y} color={color} size={size} />;
  children.push(piece);
  return children;
};
/**
 * Rerender pieces on chessboard
 * @param {*} props The properties of the component
 */
export function redrawPieces(props) {
  const { unitWidth, coord, chessMoves } = props;
  const children = [];
  chessMoves.forEach((element) => {
    const { x, y, color } = element;
    const point = coord[x][y];
    drawPiece(children, point, unitWidth, color);
  });
  this.setState({ children });
}
/**
 * Add or remove piece on chessboard
 * @param {*} nextProps Next properties of the component
 * @param {*} props Current properties of the component
 */
export function addOrRemovePiece(nextProps, props) {
  const { chessMoves: nextChessMoves, unitWidth, coord } = nextProps;
  const { chessMoves } = props;
  const { children } = this.state;
  let nextChildren;
  if (chessMoves.length < nextChessMoves.length) {
    const { x, y, color } = nextChessMoves[nextChessMoves.length - 1];
    const point = coord[x][y];
    nextChildren = drawPiece(children, point, unitWidth, color);
  } else if (chessMoves.length > nextChessMoves.length) {
    const e = nextChessMoves.length - chessMoves.length;
    nextChildren = clearPiece(children, 0, e);
  }
  this.setState({ children: nextChildren.slice() });
}
/**
 * Detect if the chess moves
 * @param {object} nextProps Next properties of the component
 * @param {object} props Current properties of the component
 */
export function detectChessMovesChange(nextProps, props) {
  const { chessMoves: nextChessMoves } = nextProps;
  const { chessMoves } = props;
  if (nextChessMoves.length !== chessMoves.length) {
    addOrRemovePiece.call(this, nextProps, props);
  }
}
/**
 * Detect if the width of container is changed
 * @param {object} prevProps The previous properties of the component
 * @param {object} props Current properties of the component
 */
export function detectWidthChange(prevProps, props) {
  const { width } = props;
  const { width: prevWidth } = prevProps;
  if (prevWidth !== width) {
    redrawPieces.call(this, props);
  }
}

export default detectChessMovesChange;
