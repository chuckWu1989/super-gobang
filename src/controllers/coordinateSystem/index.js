export function setCoordinate() {
  const { width, ratio, grid } = this.props;
  if (typeof width !== 'number') {
    return false;
  }
  const coord = new Array(grid + 2);
  const trueWidth = Math.floor(width / 15) * 15;
  const trueHeight = Math.floor((trueWidth * ratio) / 15) * 15;
  const unitWidth = trueWidth / (grid + 1);
  const unitHeight = trueHeight / (grid + 1);
  for (let x = 0; x <= grid + 1; x += 1) {
    coord[x] = new Array(grid);
    for (let y = 0; y <= grid + 1; y += 1) {
      coord[x][y] = [unitWidth * x, unitHeight * y];
    }
  }
  return this.setState({ coord, width: trueWidth, height: trueHeight });
}

export default setCoordinate;
