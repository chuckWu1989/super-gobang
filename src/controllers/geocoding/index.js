/**
 * Calculate coordinates and set into state.
 */
export function setCoordinate() {
  const { width, ratio, grid } = this.props;
  if (typeof width !== 'number') {
    return false;
  }
  const coord = new Array(grid + 2);
  const unitWidth = Math.round(width / (grid + 1));
  const unitHeight = Math.round((width * ratio) / (grid + 1));
  const trueWidth = unitWidth * (grid + 1);
  const trueHeight = unitHeight * (grid + 1);
  for (let x = 0; x <= grid + 1; x += 1) {
    coord[x] = new Array(grid);
    for (let y = 0; y <= grid + 1; y += 1) {
      coord[x][y] = [unitWidth * x, unitHeight * y];
    }
  }
  return this.setState({
    coord,
    unitWidth,
    unitHeight,
    width: trueWidth,
    height: trueHeight,
  });
}
/**
 * Detect if the width of container is changed
 * @param {*} nextProps Next properties of the component
 * @param {*} props Current properties of the component
 */
export function detectWidthChanged(nextProps, props) {
  const { width: nextWidth } = nextProps;
  const { width } = props;
  if (nextWidth !== width) {
    setCoordinate.call(this);
  }
}

export default setCoordinate;
