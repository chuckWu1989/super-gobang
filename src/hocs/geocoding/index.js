import React, { Component } from 'react';
import { setCoordinate } from '../../controllers/geocoding';

const geocoding = WrappedComponent => (
  class extends Component {
    static defaultProps = {
      grid: 15,
      ratio: 1,
      width: 400,
    };
    componentWillMount() {
      setCoordinate.call(this);
    }
    componentWillReceiveProps() {
      setCoordinate.call(this);
    }
    render() {
      const {
        coord,
        width,
        height,
        unitWidth,
        unitHeight,
      } = this.state;
      return coord ?
        (
          <WrappedComponent
            {...this.props}
            coord={coord}
            width={width}
            height={height}
            unitWidth={unitWidth}
            unitHeight={unitHeight}
          />
        ) :
        undefined;
    }
  }
);

export default geocoding;
