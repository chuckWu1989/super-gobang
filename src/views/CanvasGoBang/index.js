import React from 'react';
import CanvasChessBoard from '../../router/CanvasChessBoard';
import CanvasPieces from '../../router/CanvasPieces';
import geocoding from '../../router/geocoding';
import './style.less';

export const CanvasGoBang = props => (
  <div className="canvas-gobang-style">
    <CanvasChessBoard {...props} />
    <CanvasPieces {...props} />
  </div>
);

export default geocoding(CanvasGoBang);
