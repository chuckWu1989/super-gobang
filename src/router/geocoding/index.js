import { setCoordinate, detectWidthChanged } from '../../controllers/geocoding';
import { connect } from '../../hocs/connect';
import geocoding from '../../hocs/geocoding';

export const method = () => ({
  setCoordinate,
  detectWidthChanged,
});

export default WrapperedComponent => connect(method)(geocoding(WrapperedComponent));
