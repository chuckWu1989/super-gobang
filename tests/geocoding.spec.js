import React from 'react';
import toJson from 'enzyme-to-json';
import geocoding from '../src/hocs/geocoding';
import { CanvasGoBang } from '../src/views/CanvasGoBang';
import { method } from '../src/router/geocoding';
import * as controller from '../src/controllers/geocoding';

const unitWidth = 10;
const unitHeight = 10;
const props = {
  width: 160,
  ratio: 1,
  grid: 15,
};

describe('geocoding', () => {
  describe('view', () => {
    it('should be rendered properly', () => {
      const WrappedComponent = geocoding(CanvasGoBang);
      const mockFn = jest.fn();
      const ownProps = {
        ...props,
        setCoordinate: function coord() {
          this.setState({
            coord: [],
            unitWidth,
            unitHeight,
            height: 100,
            ...props,
          });
          mockFn();
        },
        detectWidthChanged: jest.fn(),
      };
      const wrapper = shallow(
        <WrappedComponent {...ownProps} />,
        { lifecycleExperimental: true },
      );
      const tree = toJson(wrapper);
      wrapper.setProps({ width: 200 });
      expect(tree).toMatchSnapshot();
      expect(mockFn).toHaveBeenCalledTimes(1);
      expect(ownProps.detectWidthChanged).toHaveBeenCalledTimes(1);
    });
    it('should render nothing if coord is undefined', () => {
      const WrappedComponent = geocoding(CanvasGoBang);
      const tree = toJson(shallow(<WrappedComponent {...props} />));
      expect(tree).toMatchSnapshot();
    });
  });
  describe('router', () => {
    it('should pass setCoordinate', () => {
      expect('setCoordinate' in method()).toBeTruthy();
    });
    it('should pass detectWidthChanged', () => {
      expect('detectWidthChanged' in method()).toBeTruthy();
    });
  });
  describe('controller', () => {
    let self;
    beforeEach(() => {
      self = {
        props,
        setState: function set(state) { this.state = state; },
        getState: function get() { return this.state; },
      };
    });
    describe('setCoordinate', () => {
      it('should set coordinates, unitWidth, unitHeight, width and height into state', () => {
        controller.setCoordinate.call(self);
        const state = self.getState();
        expect(state.unitWidth).toBe(10);
        expect(state.unitHeight).toBe(10);
        expect(state.width).toBe(160);
        expect(state.height).toBe(160);
        expect('coord' in state).toBeTruthy();
      });
      it('should return false if width is not a number', () => {
        const newProps = Object.assign({}, props, { width: undefined });
        self.props = newProps;
        const result = controller.setCoordinate.call(self);
        expect(result).toBeFalsy();
      });
    });
    describe('detectWidthChanged', () => {
      it('should call setCoordinate if next width is different with current width', () => {
        const nextProps = { ...props, width: 100 };
        controller.detectWidthChanged.call(self, nextProps, props);
        expect(self.state).toBeDefined();
      });
      it('should skip if next width is same with current width', () => {
        controller.detectWidthChanged.call(self, props, props);
        expect(self.state).toBeUndefined();
      });
    });
  });
});
