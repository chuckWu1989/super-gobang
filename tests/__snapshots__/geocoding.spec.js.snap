// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geocoding view should be rendered properly 1`] = `
<CanvasGoBang
  coord={Array []}
  detectWidthChanged={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "detectWidthChanged": [MockFunction] {
              "calls": [Circular],
            },
            "grid": 15,
            "ratio": 1,
            "setCoordinate": [Function],
            "width": 200,
          },
          Object {
            "detectWidthChanged": [MockFunction] {
              "calls": [Circular],
            },
            "grid": 15,
            "ratio": 1,
            "setCoordinate": [Function],
            "width": 160,
          },
        ],
      ],
    }
  }
  grid={15}
  height={100}
  ratio={1}
  setCoordinate={[Function]}
  unitHeight={10}
  unitWidth={10}
  width={160}
/>
`;
